huffman/test/bit_span_test.cpp

```diff
@@ -92,4 +92,72 @@ auto main() -> int
       huffman::bit_span{data.begin(), bit_size, bit_offset};
     }));
   } | std::vector<std::uint8_t>{8, 9, 10};  // NOLINT(readability-magic-numbers)
+
+  test("consume") = [] {
+    // NOLINTBEGIN(readability-magic-numbers)
+    static constexpr std::array data{
+        std::byte{0b10101010}, std::byte{0b01010101}};
+    huffman::bit_span span{data.data(), data.size() * CHAR_BIT};
+    expect(*span.begin() == 0_b);
+    expect(span.data() == data.data());
+    // should be a no-op now.
+    span.consume_to_byte_boundary();
+    expect(*span.begin() == 0_b);
+    expect(span.data() == data.data());
+
+    span.consume(1);
+    expect(*span.begin() == 1_b);
+    expect(span.data() == data.data());
+
+    span.consume(1);
+    expect(*span.begin() == 0_b);
+    expect(span.data() == data.data());
+
+    span.consume_to_byte_boundary();
+    expect(*span.begin() == 1_b);
+    expect(span.data() != data.data());
+    span.consume(2);
+    expect(*span.begin() == 1_b);
+    span.consume(5);
+    expect(*span.begin() == 0_b);
+
+    span.consume(1);
+    // span is now empty. Accssing .begin() would be undefined behavior.
```

Comment:
This is actually fine and you can test it out with a constant expression.

```cpp
constexpr auto bits = huffman::bit_span{nullptr, 0};
static_assert(bits.begin() == bits.end());
```

If it compiles, it's defined behavior.

However, dereferencing `begin()` is UB.